README.md
#DEMO
http://angularyii2.github.io/

# [Wiki (FAQ)](https://github.com/AngularYii2/angularyii2.github.io/wiki)


CHANGELOG.md
### HEAD

### 0.0.3

* Fixed "Error 404" ([#8](https://github.com/githubjeka/angular-yii2/issues/8))
* Update vendors ([#4](https://github.com/AngularYii2/angularyii2.github.io/pull/4)).

modules/post/controllers/PostCtrl.js
app
    .controller('PostIndex', ['$scope', 'rest', 'toaster', '$sce', 'status', '$filter', function ($scope, rest, toaster, $sce, status, $filter) {

        rest.path = 'v1/posts';

        var errorCallback = function (data) {
            toaster.clear();
            toaster.pop('error', "status: " + data.status + " " + data.name, data.message);
        };

        rest.models().success(function (data) {
            $scope.posts = $filter('filter')(data, {"status": status});
        }).error(errorCallback);

    }])

    .controller('PostView', ['$scope', 'rest', 'toaster', function ($scope, rest, toaster) {

        rest.path = 'v1/posts';

        $scope.post = {};

        var errorCallback = function (data) {
            toaster.clear();
            if (data.status == undefined) {
                angular.forEach(data, function (error) {
                    toaster.pop('error', "Field: " + error.field, error.message);
                });
            } else {
                toaster.pop('error', "code: " + data.code + " " + data.name, data.message);
            }
        };

        rest.model().success(function (data) {
            $scope.post = data;
            $scope.post.comment = {"post_id": data.id, "content": null};
        }).error(errorCallback);

        $scope.addComment = function () {

            rest.path = 'v1/comments';

            rest.postModel($scope.post.comment)
                .success(function (data) {
                    if (!angular.isArray($scope.post.comments)) {
                        $scope.post.comments = [];
                    }
                    $scope.post.comments.unshift(data);
                    $scope.post.comment.content = null;
                    toaster.pop('success', "Comment is added!");
                })
                .error(errorCallback);
        };

        $scope.deleteComment = function (comment) {

            rest.path = 'v1/comments/' + comment.id;

            rest.deleteModel()
                .success(function () {
                    angular.forEach($scope.post.comments, function (value, key) {
                        if (value.id == comment.id) {
                            $scope.post.comments.splice(key, 1);
                            toaster.pop('success', "Comment is deleted!");
                            return true;
                        }
                    });
                })
                .error(errorCallback);
        };
    }])

    .controller('PostCreate', ['$scope', 'rest', 'toaster', '$window', function ($scope, rest, toaster, $window) {

        rest.path = 'v1/posts';

        $scope.post = {};

        $scope.icons = [
            {value: '1', label: 'Draft'},
            {value: '2', label: 'Published'},
        ];

        var errorCallback = function (data) {
            toaster.clear();
            if (data.status == undefined) {
                angular.forEach(data, function (error) {
                    toaster.pop('error', "Field: " + error.field, error.message);
                });
            } else {
                toaster.pop('error', "status: " + data.status + " " + data.name, data.message);
            }
        };

        $scope.save = function () {
            rest.postModel($scope.post).success(function (data) {

                toaster.pop('success', "Saved");

                $window.setTimeout(function () {
                    $window.location = '/#!/post/' + data.id;
                }, 1000);

            }).error(errorCallback);
        };
    }])

    .controller('PostEdit', ['$scope', 'rest', 'toaster', '$window', function ($scope, rest, toaster, $window) {

        var errorCallback = function (data) {
            toaster.clear();
            if (data.status == undefined) {
                angular.forEach(data, function (error) {
                    toaster.pop('error', "Field: " + error.field, error.message);
                });
            } else {
                toaster.pop('error', "status: " + data.status + " " + data.name, data.message);
            }
        };

        rest.path = 'v1/posts';

        $scope.post = {};

        $scope.icons = [
            {value: '1', label: 'Draft'},
            {value: '2', label: 'Published'},
        ];

        rest.model().success(function (data) {
            $scope.post = data;
            $scope.post.status = $scope.post.status.toString();
        }).error(errorCallback);

        $scope.save = function () {
            rest.putModel($scope.post).success(function () {

                toaster.pop('success', "Saved");

                $window.setTimeout(function () {
                    $window.location = '/#!/post/' + $scope.post.id;
                }, 1000);
            }).error(errorCallback);
        };

    }])

    .controller('PostDelete', ['$scope', 'rest', 'toaster', '$window', function ($scope, rest, toaster, $window) {

        $scope.post = {};

        var errorCallback = function (data) {
            toaster.clear();
            if (data.status == undefined) {
                angular.forEach(data, function (error) {
                    toaster.pop('error', "Field: " + error.field, error.message);
                });
            } else {
                toaster.pop('error', "status: " + data.status + " " + data.name, data.message);
            }
        };

        rest.path = 'v1/posts';

        rest.model().success(function (data) {
            $scope.post = data;
        }).error(errorCallback);


        $scope.deleteModel = function () {

            rest.path = 'v1/posts/' + $scope.post.id;

            rest.deleteModel().success(function () {

                toaster.pop('success', "Deleted!");

                $window.setTimeout(function () {
                    $window.location = '/';
                }, 1000);

            }).error(errorCallback);
        }

    }]);

modules/site/controllers/SiteCtrl.js
app
    .controller('SiteLogin', ['$scope', 'rest', 'toaster', '$window', function ($scope, rest, toaster, $window) {

        rest.path = 'v1/user/login';

        var errorCallback = function (data) {
            toaster.clear();
            delete $window.sessionStorage._auth;
            angular.forEach(data, function (error) {
                toaster.pop('error', "Field: " + error.field, error.message);
            });
        };

        $scope.login = function () {
            rest.postModel($scope.model).success(function (data) {
                $window.sessionStorage._auth = data;
                toaster.pop('success', "Success");
                $window.setTimeout(function () {
                    $window.location = '/';
                }, 1000);
            }).error(errorCallback);
        };

    }]);

app.js
var app = angular.module('myApp', ['ngRoute', 'ngAnimate', 'toaster', 'ngSanitize', 'mgcrea.ngStrap']);

app.config(['$locationProvider', '$routeProvider', '$httpProvider', function ($locationProvider, $routeProvider, $httpProvider) {

    var modulesPath = 'modules';

    $routeProvider

        .when('/', {
            templateUrl: modulesPath + '/site/views/main.html'
        })

        .when('/login', {
            templateUrl: modulesPath + '/site/views/login.html',
            controller: 'SiteLogin'
        })

        .when('/post/published', {
            templateUrl: modulesPath + '/post/views/index.html',
            controller: 'PostIndex',
            resolve: {
                status: function () {
                    return 2;
                }
            }
        })

        .when('/post/draft', {
            templateUrl: modulesPath + '/post/views/index.html',
            controller: 'PostIndex',
            resolve: {
                status: function () {
                    return 1;
                }
            }
        })

        .when('/post/create', {
            templateUrl: modulesPath + '/post/views/form.html',
            controller: 'PostCreate'
        })

        .when('/post/:id/edit', {
            templateUrl: modulesPath + '/post/views/form.html',
            controller: 'PostEdit'
        })

        .when('/post/:id/delete', {
            templateUrl: modulesPath + '/post/views/delete.html',
            controller: 'PostDelete'
        })

        .when('/post/:id', {
            templateUrl: modulesPath + '/post/views/view.html',
            controller: 'PostView'
        })

        .when('/404', {
            templateUrl: '404.html'
        })

        .otherwise({redirectTo: '/404'})
    ;

    $locationProvider.html5Mode(true).hashPrefix('!');
    $httpProvider.interceptors.push('authInterceptor');
}]);

app.factory('authInterceptor', function ($q, $window) {
    return {
        request: function (config) {
            if ($window.sessionStorage._auth && config.url.substring(0, 4) == 'http') {
                config.params = {'access-token': $window.sessionStorage._auth};
            }
            return config;
        },
        responseError: function (rejection) {
            if (rejection.status === 401) {
                $window.setTimeout(function () {
                    $window.location = '/#!/login';
                }, 1000);
            }
            return $q.reject(rejection);
        }
    };
});

app.value('app-version', '0.0.3');

// Need set url REST Api in controller!
app.service('rest', function ($http, $location, $routeParams) {

    return {

        baseUrl: 'https://yii2-rest-githubjeka.c9.io/rest/web/',
        path: undefined,

        models: function () {
            return $http.get(this.baseUrl + this.path + location.search);
        },

        model: function () {
            if ($routeParams.expand != null) {
                return $http.get(this.baseUrl + this.path + "/" + $routeParams.id + '?expand=' + $routeParams.expand);
            }
            return $http.get(this.baseUrl + this.path + "/" + $routeParams.id);
        },

        get: function () {
            return $http.get(this.baseUrl + this.path);
        },

        postModel: function (model) {
            return $http.post(this.baseUrl + this.path, model);
        },

        putModel: function (model) {
            return $http.put(this.baseUrl + this.path + "/" + $routeParams.id, model);
        },

        deleteModel: function () {
            return $http.delete(this.baseUrl + this.path);
        }
    };

});

app
    .directive('login', ['$http', function ($http) {
        return {
            transclude: true,
            link: function (scope, element, attrs) {
                scope.isGuest = window.sessionStorage._auth == undefined;
            },

            template: '<a href="login" ng-if="isGuest">Login</a>'
        }
    }])
    .filter('checkmark', function () {
        return function (input) {
            return input ? '\u2713' : '\u2718';
        };
    });
